OUTPUT_DIRECTORY and linking against a corrosion-produced library

I was using corrosion (a fairly [recent version](https://github.com/corrosion-rs/corrosion/commit/bff871054636297221c82ccb5647399b2ce4258b)) in a subproject to build a thin wrapper around the oxipng API, and was then trying to dynamically link my main c++ project against it. I was trying to do it this way vs statically linking because my main app and test executables are pretty small, whereas oxipng is pretty large. I set `LIBRARY_OUTPUT_DIRECTORY` to `${CMAKE_BINARY_DIR}/lib` which caused the library to be put in the correct place but produced the `make` error `No rule to make target 'vendor/oxipng/oxipng_encoder.dylib', needed by 'app'. Stop.` (despite the dylib now being in `lib/`)

I created a minimal replication here: https://github.com/corrosion-rs/corrosion/compare/master...mnutt:corrosion:output-directory-linking

which just creates a `consumer` executable and links the corrosion-produced library. It produces the error:

```
[ 50%] Building CXX object CMakeFiles/consumer.dir/consumer.cpp.o
gmake[2]: *** No rule to make target 'librust_lib.a', needed by 'consumer'.  Stop.
gmake[1]: *** [CMakeFiles/Makefile2:374: CMakeFiles/consumer.dir/all] Error 2
gmake: *** [Makefile:91: all] Error 2
CMake Error at ./corrosion/test/ConfigureAndBuild.cmake:103 (message):
  Build step failed.  Exit code: `2`
```

I tried with ninja just to see what would happen and got a very similar error. I observed it on both mac/clang and linux/gcc. Apologies if I'm just using it incorrectly or if what I'm trying to do shouldn't be possible, but I'm guessing the dependency resolution step is somehow happening prior to the output_directory move? FWIW the workaround I eventually used was to statically link the corrosion-produced target against an intermediate `SHARED` target (along with some `-force_load`/`--whole-archive` stuff) and link _that_ to my main app, which worked ok.

--- a/cmake/Corrosion.cmake
+++ b/cmake/Corrosion.cmake
@@ -123,12 +123,21 @@ endfunction()
 # directory target property value at configure time. This function must be deferred to the end of
 # the configure stage, so we can be sure that the output directory is not modified afterwards.
 function(_corrosion_set_imported_location_deferred target_name base_property output_directory_property filename)
-    get_target_property(output_directory ${target_name} "${output_directory_property}")
+    # The output directory property is expected to be set on the exposed target (without postfix),
+    # but we need to set the imported location on the actual library target with postfix.
+    if("${target_name}" MATCHES "^([^-]+)-(static|shared)$")
+        set(output_dir_prop_target_name "${CMAKE_MATCH_1}")
+    else()
+        set(output_dir_prop_target_name "${target_name}")
+    endif()
+
+    get_target_property(output_directory "${output_dir_prop_target_name}" "${output_directory_property}")
+    message(DEBUG "Output directory property (target ${output_dir_prop_target_name}): ${output_directory_property} dir: ${output_directory}")
 
     if(CMAKE_CONFIGURATION_TYPES AND CMAKE_VERSION VERSION_GREATER_EQUAL 3.20.0)
         foreach(config_type ${CMAKE_CONFIGURATION_TYPES})
             string(TOUPPER "${config_type}" config_type_upper)
-            get_target_property(output_dir_curr_config ${target_name}
+            get_target_property(output_dir_curr_config "${output_dir_prop_target_name}"
                 "${output_directory_property}_${config_type_upper}"
             )
             if(output_dir_curr_config)

--- a/test/output directory/CMakeLists.txt
+++ b/test/output directory/CMakeLists.txt
@@ -45,7 +45,7 @@ add_test(NAME output_directory_bin
         -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
         "${CMAKE_CURRENT_BINARY_DIR}/build/custom_bin/${bin_name}"
 )
-set_tests_properties("output_directory_bin" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+set_tests_properties("output_directory_bin" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
 
 set(lib_name "rust_lib")
 
@@ -61,7 +61,7 @@ add_test(NAME output_directory_staticlib
         -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
         "${CMAKE_CURRENT_BINARY_DIR}/build/custom_archive/${static_lib_name}"
 )
-set_tests_properties("output_directory_staticlib" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+set_tests_properties("output_directory_staticlib" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
 
 if(WIN32)
     set(dynamic_lib_name "${lib_name}.dll")
@@ -77,7 +77,7 @@ add_test(NAME output_directory_cdylib
         -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
         "${CMAKE_CURRENT_BINARY_DIR}/build/custom_lib/${dynamic_lib_name}"
 )
-set_tests_properties("output_directory_cdylib" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+set_tests_properties("output_directory_cdylib" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
 
 if(WIN32)
     if(MSVC)
@@ -95,7 +95,7 @@ if(WIN32)
         # https://cmake.org/cmake/help/v3.25/manual/cmake-buildsystem.7.html#archive-output-artifacts
         "${CMAKE_CURRENT_BINARY_DIR}/build/custom_archive/${implib_name}"
         )
-    set_tests_properties("output_directory_implib" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+    set_tests_properties("output_directory_implib" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
 
     if(MSVC)
         set(lib_pdb_name "${lib_name}.pdb")
@@ -105,7 +105,7 @@ if(WIN32)
             -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
             "${CMAKE_CURRENT_BINARY_DIR}/build/custom_lib_pdb/${lib_pdb_name}"
             )
-        set_tests_properties("output_directory_cdylib_pdb" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+        set_tests_properties("output_directory_cdylib_pdb" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
 
         set(bin_pdb_name "rust_bin.pdb")
         add_test(NAME output_directory_bin_pdb
@@ -114,7 +114,7 @@ if(WIN32)
             -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
             "${CMAKE_CURRENT_BINARY_DIR}/build/custom_bin_pdb/${bin_pdb_name}"
             )
-        set_tests_properties("output_directory_bin_pdb" PROPERTIES FIXTURES_REQUIRED "output_directory_build")
+        set_tests_properties("output_directory_bin_pdb" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")
     endif()
 endif()
 
@@ -124,3 +124,4 @@ add_test(NAME postbuild_custom_command
     -P "${CMAKE_CURRENT_SOURCE_DIR}/TestFileExists.cmake"
     "${CMAKE_CURRENT_BINARY_DIR}/build/another_dir/moved_bin"
     )
+set_tests_properties("postbuild_custom_command" PROPERTIES FIXTURES_REQUIRED "build_fixture_output_directory")

--- a/test/output directory/output directory/CMakeLists.txt
+++ b/test/output directory/output directory/CMakeLists.txt
@@ -24,3 +24,8 @@ add_custom_command(TARGET cargo-build_rust_bin POST_BUILD
     COMMAND
     ${CMAKE_COMMAND} -E copy_if_different "$<TARGET_PROPERTY:rust_bin,LOCATION>" "${CMAKE_CURRENT_BINARY_DIR}/another_dir/moved_bin"
     )
+
+add_executable(consumer consumer.cpp)
+add_dependencies(consumer cargo-build_rust_lib)
+
+target_link_libraries(consumer rust_lib)

--- a/test/output directory/output directory/consumer.cpp
+++ b/test/output directory/output directory/consumer.cpp
@@ -1,0 +1,16 @@
+#include <iostream>
+#include <cstdlib>
+
+extern "C" unsigned int ret_12();
+
+
+int main(int argc, char *argv[])
+{
+    std::cout << "HI\n";
+    unsigned int a = ret_12();
+    if (a != 12) {
+        return -1;
+    }
+
+    return 0;
+}

